src/EventHandler.ts

```diff
@@ -0,0 +1,248 @@
+            element.addEventListener("mouseleave", this.handleMouseLeave.bind(this));
+        }
+    }
+
+    private isTouchDevice() {
+        return !!('ontouchstart' in window || navigator.maxTouchPoints);
+    }
+
+    private handleMouseEventStart = (event: MouseEvent): void => {
+        this.pendingMouseEventStart = event;
+    }
+
+    private handleTouchEventStart = (event: TouchEvent): void => {
+        if (event.changedTouches.length !== 1) {
+            // This is a multi-touch event. Ignore.
+            return;
+        }
+
+        this.pendingTouchEventStart = event;
+    }
+
+    private handleMouseEventEnd = (event: MouseEvent): void => {
+        if (!this.pendingMouseEventStart) {
+            // Somehow we got an end event without a start event. Ignore it.
+            return;
+        }
+
+        const width = window.innerWidth;
+        const height = window.innerHeight;
+
+        const relativeXDistance = (this.pendingMouseEventStart.clientX - event.clientX) / width;
+        const relativeYDistance = (this.pendingMouseEventStart.clientY - event.clientY) / height;
+
+        // Is the end event in the same place as the start event?
+        if (Math.abs(relativeXDistance) < 0.1 && Math.abs(relativeYDistance) < 0.1) {
+            if (this.pendingMouseEventEnd) {
+                // This was a double click. Let the browser handle it.
+                this.pendingMouseEventStart = null;
+                this.pendingMouseEventEnd = null;
+                return;
+            }
+
+            // This was a single click.
+            this.pendingMouseEventStart = null;
+            this.pendingMouseEventEnd = event;
+            setTimeout(this.handleClick, 200);
+            return;
+        }
+
+        this.pendingMouseEventEnd = null;
+
+        // This is a swipe or highlight. Let the browser handle it.
+        // (Swipes aren't handled on desktop.)
+        this.pendingMouseEventStart = null;
+    }
+
+    private handleTouchEventEnd = (event: TouchEvent): void => {
+        if (event.changedTouches.length !== 1) {
+            // This is a multi-touch event. Ignore.
+            return;
+        }
+
+        if (!this.pendingTouchEventStart) {
+            // Somehow we got an end event without a start event. Ignore it.
+            return;
+        }
+
+        const width = window.innerWidth;
+        const height = window.innerHeight;
+
+        const startTouch = this.pendingTouchEventStart.changedTouches[0];
+        const endTouch = event.changedTouches[0];
+
+        if (!startTouch) {
+            // Somehow we saved a touch event with no touches.
+            return;
+        }
+
+        const relativeXDistance = (startTouch.clientX - endTouch.clientX) / width;
+        const relativeYDistance = (startTouch.clientY - endTouch.clientY) / height;
+
+        // Is the end event in the same place as the start event?
+        if (Math.abs(relativeXDistance) < 0.1 && Math.abs(relativeYDistance) < 0.1) {
```

Comment:
They are similar, but I think there are enough differences that it's not worth it at this point.